File: src/main/java/io/github/seggan/jyxal/runtime/list/JyxalList.java
package io.github.seggan.jyxal.runtime.list;

import io.github.seggan.jyxal.runtime.math.BigComplex;

import java.math.BigInteger;
import java.util.AbstractList;
import java.util.Collection;
import java.util.Iterator;
import java.util.List;
import java.util.Objects;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.function.Supplier;

public abstract class JyxalList extends AbstractList<Object> {

    public static JyxalList create(Iterator<Object> generator) {
        return new InfiniteList(generator);
    }

    public static JyxalList create(Object... array) {
        return new FiniteList(List.of(array));
    }

    public static JyxalList create(Collection<?> collection) {
        return new FiniteList(collection);
    }

    public static JyxalList create() {
        return new FiniteList();
    }

    /**
     * Create an infinite list
     */
    public static JyxalList createInf(Supplier<Object> generator) {
        return new InfiniteList(new Iterator<>() {
            @Override
            public boolean hasNext() {
                return true;
            }

            @Override
            public Object next() {
                return generator.get();
            }
        });
    }

    public static JyxalList range(BigComplex start, BigComplex end) {
        return new InfiniteList(new Iterator<>() {
            BigComplex cur = start;

            @Override
            public boolean hasNext() {
                return cur.compareTo(end) < 0;
            }

            @Override
            public Object next() {
                Object ret = cur;
                cur = cur.add(BigComplex.ONE);
                return ret;
            }
        });
    }

    protected static String vyxalListFormat(List<Object> list) {
        StringBuilder sb = new StringBuilder();
        sb.append("\u27E8");
        Iterator<Object> it = list.iterator();
        while (true) {
            if (!it.hasNext()) {
                sb.delete(sb.length() - 3, sb.length());
                return sb.append("\u27E9").toString();
            }
            sb.append(it.next());
            sb.append(" | ");
        }
    }

    /**
     * Whether there exists an element at the given index
     */
    public abstract boolean hasInd(int ind);

    public abstract void mapInPlace(Function<Object, Object> f);

    public abstract void filterInPlace(Predicate<Object> p);

    public JyxalList removeAtIndex(BigInteger ind) {
        Iterator<Object> it = this.iterator();
        return new InfiniteList(new Iterator<>() {
            BigInteger i = BigInteger.ZERO;

            @Override
            public boolean hasNext() {
                return it.hasNext();
            }

            @Override
            public Object next() {
                if (i.equals(ind)) {
                    it.next();
                    i = i.add(BigInteger.ONE);
                }
                i = i.add(BigInteger.ONE);
                return it.next();
            }
        });
    }

    public JyxalList map(Function<Object, Object> f) {
        Iterator<Object> it = this.iterator();
        return new InfiniteList(new Iterator<>() {
            @Override
            public boolean hasNext() {
                return it.hasNext();
            }

            @Override
            public Object next() {
                return f.apply(it.next());
            }
        });
    }

    public JyxalList filter(Predicate<Object> pred) {
        Iterator<Object> it = this.iterator();
        return new InfiniteList(new Iterator<>() {
            /** An element waiting to be pulled */
            Object waiting = null;

            @Override
            public boolean hasNext() {
                if (waiting != null) {
                    return true;
                }
                findNext();
                return waiting != null;
            }

            @Override
            public Object next() {
                if (waiting != null) {
                    return waiting;
                }
                this.findNext();
                assert waiting != null;
                return waiting;
            }

            private void findNext() {
                while (it.hasNext()) {
                    Object next = it.next();
                    if (pred.test(next)) {
                        this.waiting = next;
                        break;
                    }
                }
            }
        });
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof JyxalList other)) return false;

        Iterator<Object> it1 = this.iterator();
        Iterator<Object> it2 = other.iterator();
        while (it1.hasNext() && it2.hasNext()) {
            if (!Objects.equals(it1.next(), it2.next())) {
                return false;
            }
        }

        return !it1.hasNext() && !it2.hasNext();
    }
}

File: src/main/kotlin/io/github/seggan/jyxal/runtime/list/JyxalList.kt
package io.github.seggan.jyxal.runtime.list

import io.github.seggan.jyxal.runtime.ProgramStack
import io.github.seggan.jyxal.runtime.jyxal
import io.github.seggan.jyxal.runtime.math.BigComplex
import io.github.seggan.jyxal.runtime.plus
import java.math.BigInteger

abstract class JyxalList : Collection<Any> {

    companion object {

        fun create(generator: Iterator<Any>): JyxalList {
            return LazyList(generator)
        }

        fun create(vararg array: Any): JyxalList {
            return FiniteList(listOf(*array))
        }

        fun create(collection: Collection<Any>): JyxalList {
            return FiniteList(collection.toList())
        }

        fun create(list: List<Any>): JyxalList {
            return FiniteList(ArrayList(list))
        }

        @JvmStatic
        @Suppress("UNNECESSARY_NOT_NULL_ASSERTION")
        fun create(stack: ProgramStack): JyxalList {
            val list = ArrayList<Any>(stack.size)
            while (!stack.isEmpty()) {
                list.add(stack.removeLast()!!)
            }
            return FiniteList(list)
        }

        fun create(): JyxalList {
            return FiniteList()
        }

        /**
         * Create an infinite list
         */
        fun createInf(generator: () -> Any): JyxalList {
            return LazyList(object : Iterator<Any> {
                override fun hasNext(): Boolean {
                    return true
                }

                override fun next(): Any {
                    return generator()
                }
            })
        }

        fun range(start: BigComplex, end: BigComplex): JyxalList {
            return LazyList(object : Iterator<BigComplex> {
                var current = start

                override fun hasNext(): Boolean {
                    return current < end
                }

                override fun next(): BigComplex {
                    val result = current
                    current += 1
                    return result
                }
            })
        }

        fun range(start: Int, end: Int): JyxalList {
            return LazyList(object : Iterator<BigComplex> {
                var current = start

                override fun hasNext(): Boolean {
                    return current < end
                }

                override fun next(): BigComplex {
                    val result = current++
                    return result.jyxal()
                }
            })
        }

        fun fromIterableLazy(iterable: Iterable<Any>): JyxalList {
            return LazyList(iterable.iterator())
        }
    }

    protected fun vyxalListFormat(list: List<Any>): String {
        val sb = StringBuilder()
        sb.append("⟨")
        val it = list.iterator()
        while (true) {
            if (!it.hasNext()) {
                if (sb.length > 1) {
                    sb.delete(sb.length - 3, sb.length)
                }
                return sb.append("⟩").toString()
            }
            sb.append(it.next())
            sb.append(" | ")
        }
    }

    /**
     * Whether there exists an element at the given index
     */
    abstract fun hasInd(ind: Int): Boolean

    abstract fun isLazy(): Boolean

    abstract fun toNonLazy(): JyxalList

    abstract operator fun get(ind: Int): Any
    abstract operator fun get(ind: IntProgression): JyxalList

    abstract fun hasAtLeast(amount: Int): Boolean

    abstract fun listIterator(): ListIterator<Any>

    override fun iterator(): Iterator<Any> {
        return listIterator()
    }

    abstract fun remove(ind: Int): JyxalList

    abstract fun map(f: (Any) -> Any): JyxalList

    abstract fun filter(pred: (Any) -> Boolean): JyxalList

    abstract fun add(ind: BigInteger, value: Any): JyxalList

    abstract fun add(value: Any): JyxalList

    abstract fun addAll(iterable: Iterable<Any>): JyxalList

    abstract fun zip(iterable: Iterable<Any>): JyxalList

    abstract fun zip(iterable: Iterable<Any>, f: (Any, Any) -> Any): JyxalList

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is JyxalList) return false

        val it1 = this.iterator()
        val it2 = other.iterator()
        while (it1.hasNext() && it2.hasNext()) {
            if (it1.next() != it2.next()) {
                return false
            }
        }

        return !it1.hasNext() && !it2.hasNext()
    }

    abstract override fun hashCode(): Int

    override fun containsAll(elements: Collection<Any>): Boolean {
        for (element in elements) {
            if (!this.contains(element)) {
                return false
            }
        }
        return true
    }

    override fun isEmpty(): Boolean {
        return !this.iterator().hasNext()
    }
}
